compiler/rustc_monomorphize/src/partitioning/mod.rs

```diff
@@ -207,15 +207,19 @@ pub fn partition<'tcx>(
     if instrument_dead_code {
         // Find the smallest CGU that has exported symbols and put the dead
         // function stubs in that CGU. We look for exported symbols to increase
-        // the likelyhood the linker won't throw away the dead functions.
-        let mut cgus_with_exported_symbols: Vec<_> = post_inlining
-            .codegen_units
-            .iter_mut()
+        // the likelihood the linker won't throw away the dead functions.
+        // FIXME(#92165): In order to truly resolve this, we need to make sure
+        // the object file (CGU) containing the dead function stubs is included
+        // in the final binary. This will probably require forcing these
+        // function symbols to be included via `-u` or `/include` linker args.
+        let mut cgus: Vec<_> = post_inlining.codegen_units.iter_mut().collect();
+        cgus.sort_by_key(|cgu| cgu.size_estimate());
+
+        let dead_code_cgu = cgus
+            .into_iter()
             .filter(|cgu| cgu.items().iter().any(|(_, (linkage, _))| *linkage == Linkage::External))
-            .collect();
-        cgus_with_exported_symbols.sort_by_key(|cgu| cgu.size_estimate());
-
-        let dead_code_cgu = cgus_with_exported_symbols.last_mut().unwrap();
+            .last()
+            .unwrap();
```

Comment:
Is it possible we'd have no CGUs with externally linked symbols?

Reply:
If I understand correctly, a CGU compiles an LLVM Module.

Logically (but maybe naively), I assume a Module without any externally linked symbols could not be incorporated into a program that you could then link and run. And if that's the case, there would be no way to run a binary, using that Module, in order to generate coverage profraw data.

So even if we had a Module with no externally linked symbols, the only way to add coverage mapping data (and unused function placeholders) would be to add them to a Module that is not being compiled (which is, at least, impractical).

Reply:
Yeah, @richkadel is correct. There are basically two cases: we are compiling an executable or we are compiling a library. For executables, the entry point is externally visible/linked. For libraries, something has to be externally linked or the library isn't useable.

Reply:
> For libraries, something has to be externally linked or the library isn't useable.

Ah, so this isn't quite true. If a library contains only macros, then it will have one CGU which is empty. In this case, I'll just fall back to the first CGU and also add an assert to verify there is at least one CGU.